# CS 109A/STAT 121A/AC 209A/CSCI E-109A: Homework 8
# Ensemble methods

**Harvard University**<br/>
**Fall 2017**<br/>
**Instructors**: Pavlos Protopapas, Kevin Rader, Rahul Dave


Import libraries:

In [1]:
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import sklearn.metrics as metrics
from sklearn.model_selection import cross_val_score
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.linear_model import LogisticRegressionCV
%matplotlib inline

# Higgs Boson Discovery

The discovery of the Higgs boson in July 2012 marked a fundamental breakthrough in particle physics. The Higgs boson particle was discovered through experiments at the Large Hadron Collider at CERN, by colliding beams of protons at high energy. A key challenge in analyzing the results of these experiments is to differentiate between a collision that produces Higgs bosons and collisions thats produce only background noise. We shall explore the use of ensemble methods for this classification task.

You are provided with data from Monte-Carlo simulations of collisions of particles in a particle collider experiment. The training set is available in `Higgs_train.csv` and the test set is in `Higgs_test.csv`. Each row in these files corresponds to a particle colision described by 28 features (columns 1-28), of which the first 21 features are kinematic properties measured by the particle detectors in the accelerator, and the remaining features are derived by physicists from the the first 21 features. The class label is provided in the last column, with a label of 1 indicating that the collision produces Higgs bosons (signal), and a label of 0 indicating that the collision produces other particles (background). 

The data set provided to you is a small subset of the HIGGS data set in the UCI machine learning repository. The following paper contains further details about the data set and the predictors used: <a href = "https://www.nature.com/articles/ncomms5308">Baldi et al., Nature Communications 5, 2014</a>.


## Question 1 (2pt): Single Decision Tree
We start by building a basic model which we will use as our base model for comparison. 

1. Fit a decision tree model to the training set and report the classification accuracy of the model on the test set. Use 5-fold cross-validation to choose the (maximum) depth for the tree. You will use the max_depth you find here throughout the homework. 


In [2]:
train = pd.read_csv("Higgs_train.csv")
train.columns = train.columns.map(lambda x: x.replace(' ', '_'))
train.head()

lepton_pT  _lepton_eta  _lepton_phi  _missing_energy_magnitude  \
0   0.376816    -1.583727    -1.707552                   0.990897   
1   0.707330     0.087603    -0.399742                   0.918742   
2   0.617290     0.265839    -1.345227                   1.154581   
3   0.850992    -0.380876    -0.071264                   1.468704   
4   0.767540    -0.691572    -0.040191                   0.614843   

   _missing_energy_phi  _jet_1_pt  _jet_1_eta  _jet_1_phi  _jet_1_b-tag  \
0             0.114397   1.253553    0.619859   -1.479572      2.173076   
1            -1.229936   1.172847   -0.552574    0.886053      2.173076   
2             1.036646   0.954822    0.377252   -0.147960      0.000000   
3            -0.795133   0.691818    0.883260    0.496881      0.000000   
4             0.143765   0.748614    0.397057   -0.873640      0.000000   

   _jet_2_pt   ...    _jet_4_phi  _jet_4_b-tag     _m_jj    _m_jjj     _m_lv  \
0   0.753658   ...      0.397156      0.000000  0.522449  1.318622  0.982398   
1   1.298317   ...      0.236231      0.000000  0.439696  0.828885  0.992241   
2   1.063507   ...     -0.542413      0.000000  1.024506  1.026255  0.986289   
3   1.616349   ...     -1.520171      0.000000  1.197755  1.100534  0.987262   
4   1.147862   ...      0.502034      1.550981  0.921948  0.864080  0.982839   

     _m_jlv     _m_bb    _m_wbb   _m_wwbb  _class  
0  1.359610  0.964809  1.309991  1.083203     1.0  
1  1.157820  2.215780  1.189586  0.937976     1.0  
2  0.927720  1.371080  0.981672  0.917436     1.0  
3  1.353453  1.455383  0.994682  0.953553     1.0  
4  1.373222  0.601492  0.918621  0.957063     0.0  

[5 rows x 29 columns]

In [3]:
test = pd.read_csv("Higgs_test.csv")
test.columns = test.columns.map(lambda x: x.replace(' ', '_'))
test.head()

lepton_pT  _lepton_eta  _lepton_phi  _missing_energy_magnitude  \
0   0.883751    -0.461715     0.196283                   1.330217   
1   0.780168    -0.292245     0.897072                   0.384474   
2   0.352659    -1.073368    -1.741953                   1.174700   
3   0.757292     0.821975    -1.290851                   0.207558   
4   2.038721     2.025801    -0.471873                   0.423674   

   _missing_energy_phi  _jet_1_pt  _jet_1_eta  _jet_1_phi  _jet_1_b-tag  \
0             1.522737   1.039467   -1.519038   -1.457952      2.173076   
1             0.412700   1.215078   -0.466424   -0.920208      0.000000   
2            -0.198694   0.557980    0.057408   -1.491768      1.086538   
3            -0.150971   1.222224   -1.638856    1.531346      0.000000   
4            -1.497848   1.062186    0.798100    1.218678      2.173076   

   _jet_2_pt   ...    _jet_4_phi  _jet_4_b-tag     _m_jj    _m_jjj     _m_lv  \
0   0.361469   ...     -1.137282      0.000000  1.454688  0.790714  1.397708   
1   1.104803   ...     -0.091825      0.000000  0.812113  0.727805  0.975326   
2   0.911667   ...      1.545824      0.000000  0.829461  1.060325  0.992326   
3   1.683582   ...      0.116925      3.101961  4.290139  2.415188  0.994889   
4   0.805279   ...     -0.079062      3.101961  0.894990  0.936199  1.027161   

     _m_jlv     _m_bb    _m_wbb   _m_wwbb  _class  
0  1.250985  0.712875  0.811940  0.820693     1.0  
1  0.636811  0.569234  0.776607  0.715494     1.0  
2  0.824898  0.365448  0.800015  0.765989     0.0  
3  0.923447  0.927035  1.755831  1.362970     1.0  
4  1.559567  1.148236  1.115536  1.157044     1.0  

[5 rows x 29 columns]

In [4]:
depths = [i for i in range(2, 11)]
train_scores = []
test_scores = []
for depth in depths:
    clf = DecisionTreeClassifier(max_depth = depth)
    clf.fit(train.iloc[:,:-1],train.iloc[:,-1])
    train_scores.append(clf.score(train.iloc[:,:-1],train.iloc[:,-1]))
    test_scores.append(clf.score(test.iloc[:,:-1],test.iloc[:,-1]))
print("For depth of 2, 3, ..., 10:")
print("train scores are: \n", train_scores)
print("test scores are: \n", test_scores)

For depth of 2, 3, ..., 10:
train scores are: 
 [0.63859999999999995, 0.64219999999999999, 0.66020000000000001, 0.68200000000000005, 0.71060000000000001, 0.74280000000000002, 0.7752, 0.81579999999999997, 0.8468]
test scores are: 
 [0.64300000000000002, 0.64319999999999999, 0.64959999999999996, 0.64559999999999995, 0.65620000000000001, 0.64759999999999995, 0.62739999999999996, 0.62160000000000004, 0.62539999999999996]


In [5]:
total_scores = []
for depth in depths:
    clf = DecisionTreeClassifier(max_depth = depth)  
    # Perform 5-fold cross validation 
    score = cross_val_score(estimator=clf, X=train.iloc[:,:-1], y=train.iloc[:,-1], cv=5)
    total_scores.append((depth, np.mean(score)))
total_scores

[(2, 0.62320749580749579),
 (3, 0.62041149381149385),
 (4, 0.63580570720570717),
 (5, 0.64001170881170888),
 (6, 0.62680570360570353),
 (7, 0.629007704007704),
 (8, 0.62720070060070054),
 (9, 0.62540750100750098),
 (10, 0.61760429580429577)]

In [9]:
# max(total_scores, key=lambda x:x[1])
best_depth = max(total_scores, key=lambda x:x[1])[0]
print("The depth of the tree using 5-fold cross-validation is:", best_depth)

The depth of the tree using 5-fold cross-validation is: 5


## Question 2 (15pt): Dropout-based Approach
We start with a simple method inspired from the idea of 'dropout' in machine learning, where we fit multiple decision trees on random subsets of predictors, and combine them through a majority vote. The procedure is described below.

- For each predictor in the training sample, set the predictor values to 0 with probability $p$  (i.e. drop the predictor by setting it to 0). Repeat this for $B$ trials to create $B$ separate training sets.


- Fit decision tree models $\hat{h}^1(x), \ldots, \hat{h}^B(x) \in \{0,1\}$ to the $B$ training sets. You may allow the trees to have unrestricted depth.

- Combine the decision tree models into a single classifier by taking a majority vote:
$$
\hat{H}_{maj}(x) \,=\, majority\Big(\hat{h}^1(x), \ldots, \hat{h}^B(x)\Big).
$$


We shall refer to the combined classifier as an ** *ensemble classifier* **. Implement the described dropout approach, and answer the following questions:
1. Apply the dropout procedure with $p = 0.5$ for different number of trees (say $2, 4, 8, 16, \ldots, 256$), and evaluate the training and test accuracy of the combined classifier. Does an increase in the number of trees improve the training and test performance? Explain your observations in terms of the bias-variance trade-off for the classifier.
- Fix the number of trees to 64 and apply the dropout procedure with different dropout rates $p = 0.1, 0.3, 0.5, 0.7, 0.9$. Based on your results, explain how the dropout rate influences the bias and variance of the combined classifier.
- Apply 5-fold cross-validation to choose the optimal combination of the dropout rate and number of trees. How does the test performance of an ensemble of trees fitted with the optimal dropout rate and number of trees compare with the single decision tree model in Question 1?
[hint: Training with large number of trees can take long time. You may need to restrict the max number of trees.]

#### Question 2.1: Apply the dropout procedure with $p = 0.5$ for different number of trees (say $2, 4, 8, 16, \ldots, 256$), and evaluate the training and test accuracy of the combined classifier. Does an increase in the number of trees improve the training and test performance? Explain your observations in terms of the bias-variance trade-off for the classifier.

#### Question 2.2: Fix the number of trees to 64 and apply the dropout procedure with different dropout rates $p = 0.1, 0.3, 0.5, 0.7, 0.9$. Based on your results, explain how the dropout rate influences the bias and variance of the combined classifier.

#### Question 2.3: Apply 5-fold cross-validation to choose the optimal combination of the dropout rate and number of trees. How does the test performance of an ensemble of trees fitted with the optimal dropout rate and number of trees compare with the single decision tree model in Question 1?

## Question 3 (15pt): Random Forests

We now move to a more sophisticated ensemble technique, namely random forest:
1. How does a random forest approach differ from the dropout procedure described in Question 2? 
 
- Fit random forest models to the training set for different number of trees (say $2, 4, 8, 16, \ldots, 256$), and evaluate the training and test accuracies of the models. You may set the number of predictors for each tree in the random forest model to $\sqrt{p}$, where $p$ is the total number of predictors. 

- Based on your results, do you find that a larger number of trees necessarily improves the test accuracy of a random forest model? Explain how the number of trees effects the training and test accuracy of a random forest classifier, and how this relates to the bias-variance trade-off for the classifier. 
  
- Fixing the number of trees to a reasonable value, apply 5-fold cross-validation to choose the optimal value for the  number of predictors. How does the test performance of random forest model fitted with the optimal number of trees compare with the dropout approach in Question 2?  





## Question 4 (15pt): Boosting

We next compare the random forest model with the approach of boosting:


1. Apply the AdaBoost algorithm to fit an ensemble of decision trees. Set the learning rate to 0.05, and try out different tree depths for the base learners: 1, 2, 10, and unrestricted depth.  Make a plot of the training accuracy of the ensemble classifier as a function of tree depths. Make a similar plot of the test accuracies as a function of number of trees (say $2, 4, 8, 16, \ldots, 256$).
- How does the number of trees influence the training and test performance? Compare and contrast between the trends you see in the training and test performance of AdaBoost and that of the random forest models in Question 3. Give an explanation for your observations.
- How does the tree depth of the base learner impact the training and test performance? Recall that with random forests, we allow the depth of the individual trees to be unrestricted. Would you recommend the same strategy for boosting? Explain your answer.
- Apply 5-fold cross-validation to choose the optimal number of trees $B$ for the ensemble and the optimal tree depth for the base learners. How does an ensemble classifier fitted with the optimal number of trees and the optimal tree depth compare with the random forest model fitted in Question 3.4? 



## Question 5 (3pt): Meta-classifier

We have so far explored techniques that grow a collection of trees either by creating multiple copies of the original training set, or through a sequential procedure, and then combines these trees into a single classifier. Consider an alternate scenario where you are provided with a pre-trained collection of trees, say from different participants of a data science competition for Higgs boson discovery. What would be a good strategy to combine these pre-fitted trees into a single powerful classifier? Of course, a simple approach would be to take the majority vote from the individual trees. Can we do better than this simple combination strategy?

A collection of 100 decision tree classifiers is provided in the file `models.npy` and can be loaded into an array by executing:

`models = np.load('models.npy')`

You can make predictions using the $i^\text{th}$ model on an array of predictors `x` by executing:

`model[i].predict(x)`  &nbsp;&nbsp;&nbsp;
or &nbsp;&nbsp;&nbsp;
`model[i].predict_proba(x)`

and score the model on predictors `x` and labels `y` by using:

`model[i].score(x, y)`.

1. Implement a strategy to combine the provided decision tree classifiers, and compare the test perfomance of your approach with the majority vote classifier. Explain your strategy/algorithm.



--- 

---

## APCOMP209a - Homework Question
​
We've worked with imputation methods on missing data in Homework 6.  We've worked with Decision Trees in HW7 and here.  Now let's see what happens if we try to work with Decision Trees and Missing Data at the same time! We'll be working with a dataset from the UCI Machine Learning Repository that uses a variety of wine chemical predictors to classify wines grown in the same region in Italy.  Each line represents 13 (mostly chemical) predictors of the response variable wine class, including things like alcohol content, hue , and phenols.  Unfortunately some of the predictor values were lost in measurement. Please load `wine_quality_missing.csv`. 
​
*Note*: As in HW6 be careful of reading/treating column names and row names in this data set.
​
​
1. Remove all observations that contain and missing values, split the dataset into a 75-25 train-test split, and fit the sklearn DecisionTreeClassifier and RandomForestClassifier.   Use cross-validation to find the optimal tree depth for each method.  Report the optimal tree-depth, overall classification rate and confusion matrix on the test set for each method.
2. Restart with a fresh copy of the data and impute the missing data via mean imputation.  Split the data 75-25 and again fit DecisionTreeClassifier and RandomForestClassifier using cross-validation to find the optimal tree depth.  Report the optimal tree depth, overall classification rate and confusion matrix on the test set for each method.  
3. Again restart with a fresh copy of the data but this time let's try something different.  As discussed in section, CART Decision Trees can take advantage of surrogate splits to handle missing data.  Split the data 75-25 and construct a **custom** decision tree model and train it on the training set with missing data. Report the optimal tree depth, overall classification rate and confusion matrix on the test set and compare your results to the Imputation and DecisionTree model results in part 1 & 2.
